File: server/LingoHop.Application/Decks/UseCases/ResetDeckProgressUseCase.cs
using LingoHop.Application.Abstractions;
using LingoHop.Application.Common;
using LingoHop.Application.Decks.Dtos;
using LingoHop.Domain.Decks;

namespace LingoHop.Application.Decks.UseCases;

/// <summary>Clears the mastery counters of every card so the deck can be learned again.</summary>
public sealed class ResetDeckProgressUseCase(
    ICurrentLearner currentLearner,
    IDeckRepository decks,
    IUnitOfWork unitOfWork)
{
    public async Task<Result<DeckDetailsDto>> ExecuteAsync(
        Guid deckId,
        CancellationToken cancellationToken = default)
    {
        var user = await currentLearner.GetAsync(cancellationToken);
        var deck = await decks.FindByIdAsync(deckId, cancellationToken);
        if (deck is null || !deck.IsOwnedBy(user.Id))
        {
            return DeckErrors.NotFound(deckId);
        }

        deck.ResetProgress();
        await unitOfWork.SaveChangesAsync(cancellationToken);

        return deck.ToDetailsDto();
    }
}

File: client/src/app/application/decks/reset-deck-progress.use-case.ts
import { Injectable, inject } from '@angular/core';
import { firstValueFrom } from 'rxjs';
import { DeckDetails } from '../../domain/models/deck.model';
import { DeckRepository } from '../../domain/ports/deck.repository';

/** Clears every card's mastery counters so the deck can be learned from scratch. */
@Injectable({ providedIn: 'root' })
export class ResetDeckProgressUseCase {
  private readonly decks = inject(DeckRepository);

  execute(deckId: string): Promise<DeckDetails> {
    return firstValueFrom(this.decks.resetProgress(deckId));
  }
}
